src/System/Taffybar/Widgets/PollingBar.hs

```diff
@@ -7,27 +7,27 @@ module System.Taffybar.Widgets.PollingBar (
   BarDirection(..),
   -- * Constructors and accessors
   pollingBarNew,
+  verticalBarFromCallback,
   defaultBarConfig
   ) where
 
 import Control.Concurrent
-import qualified Control.Exception.Enclosed as E
+import Control.Exception.Enclosed ( tryAny )
 import Control.Monad ( forever )
 import Graphics.UI.Gtk
+import System.Taffybar.Widgets.Util ( backgroundLoop, drawOn )
 
 import System.Taffybar.Widgets.VerticalBar
 
-pollingBarNew :: BarConfig -> Double -> IO Double -> IO Widget
-pollingBarNew cfg pollSeconds action = do
+verticalBarFromCallback :: BarConfig -> IO Double -> IO Widget
+verticalBarFromCallback cfg action = do
   (drawArea, h) <- verticalBarNew cfg
+  drawOn drawArea $
+    backgroundLoop $ do
+      esample <- tryAny action
+      traverse (verticalBarSetPercent h) esample
 
-  _ <- on drawArea realize $ do
-    _ <- forkIO $ forever $ do
-      esample <- E.tryAny action
-      case esample of
-        Left _ -> return ()
-        Right sample -> verticalBarSetPercent h sample
-      threadDelay $ floor (pollSeconds * 1000000)
-    return ()
-
-  return drawArea
+pollingBarNew :: BarConfig -> Double -> IO Double -> IO Widget
+pollingBarNew cfg pollSeconds action =
+  verticalBarFromCallback cfg $ action <* delay
```

Comment:
Actually, this introduces a bit of a change - because the threadDelay is now inside of the action, it needs to complete before verticalBarSetPercent is called. The net effect is that your bar can take up to pollSeconds to initally draw when taffybar is launched, but other than that I don't think it'd have any effect. I could try thinking of something else if that's not acceptable.
Also, I don't think `action <* delay` vs `delay *> action` really matters here.

Reply:
Perhaps splitting up the `action` that `verticalBarFromCallback` takes into two IO values - one of which blocks, and the other of which produces a value, so it can immediately paint the initial value, and then go into its blocking / polling loop - rather than having them combined into one with `*>` or `do` or whatever.

Reply:
I tried a refactor like that locally:
```haskell
do
  action >>= verticalBarSetPercent h
  wait
```
And it doesn't seem to have any effect - it still waits until the first time I adjust the brightness to display a value on the brightness graph.